File: adapters/frontend/flutter/lib/src/rampart_client.dart
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:http/http.dart' as http;
import 'package:url_launcher/url_launcher.dart';

import 'rampart_claims.dart';
import 'rampart_error.dart';

/// Configuration for [RampartClient].
class RampartClientConfig {
  /// Rampart server URL (e.g. "http://localhost:8080").
  final String issuer;

  /// OAuth 2.0 client ID registered with the Rampart server.
  final String clientId;

  /// OAuth 2.0 redirect URI -- must exactly match a registered redirect URI.
  /// For mobile, use a custom scheme (e.g. "com.example.app://callback").
  /// For web, use a standard HTTPS URL.
  final String redirectUri;

  /// OAuth 2.0 scopes (default: "openid").
  final String scope;

  /// Optional HTTP client for testing or custom configuration.
  final http.Client? httpClient;

  const RampartClientConfig({
    required this.issuer,
    required this.clientId,
    required this.redirectUri,
    this.scope = 'openid',
    this.httpClient,
  });
}

/// Token set returned by login and refresh operations.
class RampartTokens {
  final String accessToken;
  final String refreshToken;
  final String tokenType;
  final int expiresIn;

  const RampartTokens({
    required this.accessToken,
    required this.refreshToken,
    required this.tokenType,
    required this.expiresIn,
  });

  factory RampartTokens.fromJson(Map<String, dynamic> json) {
    return RampartTokens(
      accessToken: json['access_token'] as String,
      refreshToken: json['refresh_token'] as String,
      tokenType: json['token_type'] as String? ?? 'Bearer',
      expiresIn: json['expires_in'] as int? ?? 3600,
    );
  }

  Map<String, dynamic> toJson() => {
        'access_token': accessToken,
        'refresh_token': refreshToken,
        'token_type': tokenType,
        'expires_in': expiresIn,
      };
}

/// Flutter authentication client for Rampart IAM.
///
/// Implements OAuth 2.0 Authorization Code flow with PKCE.
/// Supports mobile (iOS/Android) via custom URI schemes and web via standard
/// redirect URLs. Tokens are persisted using flutter_secure_storage.
class RampartClient {
  final String _issuer;
  final String _clientId;
  final String _redirectUri;
  final String _scope;
  final http.Client _httpClient;
  final FlutterSecureStorage _storage;

  RampartTokens? _tokens;

  static const _storageKeyTokens = 'rampart_tokens';
  static const _storageKeyCodeVerifier = 'rampart_pkce_code_verifier';
  static const _storageKeyState = 'rampart_pkce_state';

  /// Creates a new [RampartClient].
  ///
  /// Call [loadStoredTokens] after construction to restore any previously
  /// persisted session.
  RampartClient(RampartClientConfig config)
      : _issuer = config.issuer.replaceAll(RegExp(r'/+$'), ''),
        _clientId = config.clientId,
        _redirectUri = config.redirectUri,
        _scope = config.scope,
        _httpClient = config.httpClient ?? http.Client(),
        _storage = const FlutterSecureStorage();

  /// Whether the client holds a non-expired access token.
  bool get isAuthenticated {
    final token = _tokens?.accessToken;
    if (token == null) return false;
    try {
      final parts = token.split('.');
      if (parts.length != 3) return false;
      // Normalize base64url to base64.
      final payload = _decodeBase64Url(parts[1]);
      final claims = jsonDecode(payload) as Map<String, dynamic>;
      final exp = claims['exp'] as int?;
      if (exp == null) return false;
      return DateTime.fromMillisecondsSinceEpoch(exp * 1000)
          .isAfter(DateTime.now());
    } catch (_) {
      return false;
    }
  }

  /// Returns the current access token, or null.
  String? get accessToken => _tokens?.accessToken;

  /// Returns a copy of the current tokens, or null.
  RampartTokens? get tokens => _tokens;

  /// Load tokens from secure storage. Call this once at app startup to
  /// restore a previous session.
  Future<void> loadStoredTokens() async {
    final stored = await _storage.read(key: _storageKeyTokens);
    if (stored != null) {
      try {
        _tokens = RampartTokens.fromJson(
          jsonDecode(stored) as Map<String, dynamic>,
        );
      } catch (_) {
        await _storage.delete(key: _storageKeyTokens);
      }
    }
  }

  /// Start the OAuth login flow by opening the authorization URL in the
  /// system browser.
  ///
  /// On mobile, the browser will redirect back to your app via the custom
  /// URI scheme configured in [RampartClientConfig.redirectUri].
  /// On web, the page will navigate to the redirect URL.
  ///
  /// After the redirect, call [handleCallback] with the full callback URI.
  Future<void> loginWithRedirect() async {
    final codeVerifier = _generateCodeVerifier();
    final codeChallenge = _generateCodeChallenge(codeVerifier);
    final state = _generateState();

    // Persist PKCE values so they survive the browser redirect.
    await _storage.write(key: _storageKeyCodeVerifier, value: codeVerifier);
    await _storage.write(key: _storageKeyState, value: state);

    final params = {
      'response_type': 'code',
      'client_id': _clientId,
      'redirect_uri': _redirectUri,
      'scope': _scope,
      'state': state,
      'code_challenge': codeChallenge,
      'code_challenge_method': 'S256',
    };

    final authorizeUrl =
        Uri.parse('$_issuer/oauth/authorize').replace(queryParameters: params);

    if (!await launchUrl(authorizeUrl, mode: LaunchMode.externalApplication)) {
      throw const RampartError(
        error: 'launch_failed',
        errorDescription: 'Could not open the authorization URL in a browser.',
      );
    }
  }

  /// Handle the OAuth callback after the browser redirects back to the app.
  ///
  /// Pass the full callback URI (including query parameters). This method
  /// validates the state parameter, exchanges the authorization code for
  /// tokens, and persists them in secure storage.
  Future<RampartTokens> handleCallback(Uri uri) async {
    final code = uri.queryParameters['code'];
    final state = uri.queryParameters['state'];
    final error = uri.queryParameters['error'];

    if (error != null) {
      final description =
          uri.queryParameters['error_description'] ?? error;
      throw RampartError(error: error, errorDescription: description);
    }

    if (code == null || state == null) {
      throw const RampartError(
        error: 'invalid_callback',
        errorDescription: 'Missing code or state parameter in callback URL.',
      );
    }

    // Validate state.
    final storedState = await _storage.read(key: _storageKeyState);
    if (storedState == null || storedState != state) {
      await _cleanupPkceStorage();
      throw const RampartError(
        error: 'state_mismatch',
        errorDescription:
            'State parameter does not match. Possible CSRF attack.',
      );
    }

    final codeVerifier = await _storage.read(key: _storageKeyCodeVerifier);
    if (codeVerifier == null) {
      await _cleanupPkceStorage();
      throw const RampartError(
        error: 'missing_verifier',
        errorDescription: 'Code verifier not found in secure storage.',
      );
    }

    // Exchange code for tokens.
    final body = {
      'grant_type': 'authorization_code',
      'code': code,
      'client_id': _clientId,
      'redirect_uri': _redirectUri,
      'code_verifier': codeVerifier,
    };

    final response = await _httpClient.post(
      Uri.parse('$_issuer/oauth/token'),
      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
      body: body,
    );

    await _cleanupPkceStorage();

    if (response.statusCode != 200) {
      throw _parseError(response);
    }

    final tokens = RampartTokens.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
    await _setTokens(tokens);
    return tokens;
  }

  /// Fetch the authenticated user profile from the /me endpoint.
  Future<RampartClaims> getUser() async {
    final response = await authFetch(Uri.parse('$_issuer/me'));

    if (response.statusCode != 200) {
      throw _parseError(response);
    }

    return RampartClaims.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
  }

  /// Make an authenticated HTTP GET request with automatic Bearer token.
  ///
  /// On a 401 response, attempts one silent token refresh and retries.
  /// Pass [method], [headers], and [body] for non-GET requests.
  Future<http.Response> authFetch(
    Uri url, {
    String method = 'GET',
    Map<String, String>? headers,
    Object? body,
  }) async {
    Future<http.Response> doFetch() {
      final mergedHeaders = {
        if (headers != null) ...headers,
        'Authorization': 'Bearer ${_tokens?.accessToken}',
      };

      final request = http.Request(method, url)..headers.addAll(mergedHeaders);
      if (body is String) {
        request.body = body;
      } else if (body is Map) {
        request.bodyFields = body.cast<String, String>();
      }

      return _httpClient.send(request).then(http.Response.fromStream);
    }

    var response = await doFetch();

    if (response.statusCode == 401 && _tokens?.refreshToken != null) {
      try {
        await refresh();
        response = await doFetch();
      } catch (_) {
        // Refresh failed -- return the original 401.
      }
    }

    return response;
  }

  /// Refresh the access token using the stored refresh token.
  Future<RampartTokens> refresh() async {
    final refreshToken = _tokens?.refreshToken;
    if (refreshToken == null) {
      throw const RampartError(
        error: 'no_refresh_token',
        errorDescription: 'No refresh token available.',
      );
    }

    final response = await _httpClient.post(
      Uri.parse('$_issuer/token/refresh'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'refresh_token': refreshToken}),
    );

    if (response.statusCode != 200) {
      await _setTokens(null);
      throw _parseError(response);
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    final updated = RampartTokens(
      accessToken: data['access_token'] as String,
      refreshToken: refreshToken,
      tokenType: data['token_type'] as String? ?? 'Bearer',
      expiresIn: data['expires_in'] as int? ?? 3600,
    );
    await _setTokens(updated);
    return updated;
  }

  /// Logout -- invalidates the refresh token on the server and clears local
  /// tokens.
  Future<void> logout() async {
    final refreshToken = _tokens?.refreshToken;
    if (refreshToken != null) {
      try {
        await _httpClient.post(
          Uri.parse('$_issuer/logout'),
          headers: {'Content-Type': 'application/json'},
          body: jsonEncode({'refresh_token': refreshToken}),
        );
      } catch (_) {
        // Best-effort server logout.
      }
    }

    await _setTokens(null);
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  Future<void> _setTokens(RampartTokens? tokens) async {
    _tokens = tokens;
    if (tokens != null) {
      await _storage.write(
        key: _storageKeyTokens,
        value: jsonEncode(tokens.toJson()),
      );
    } else {
      await _storage.delete(key: _storageKeyTokens);
    }
  }

  Future<void> _cleanupPkceStorage() async {
    await _storage.delete(key: _storageKeyCodeVerifier);
    await _storage.delete(key: _storageKeyState);
  }

  RampartError _parseError(http.Response response) {
    try {
      final json = jsonDecode(response.body) as Map<String, dynamic>;
      return RampartError.fromJson(json, status: response.statusCode);
    } catch (_) {
      return RampartError(
        error: 'unknown_error',
        errorDescription:
            'HTTP ${response.statusCode}: ${response.reasonPhrase}',
        status: response.statusCode,
      );
    }
  }

  /// Generate a random 64-character code verifier (base64url-encoded).
  String _generateCodeVerifier() {
    final random = Random.secure();
    final bytes = List<int>.generate(48, (_) => random.nextInt(256));
    return _base64UrlEncode(bytes);
  }

  /// Compute S256 code challenge: BASE64URL(SHA256(verifier)).
  String _generateCodeChallenge(String verifier) {
    final bytes = utf8.encode(verifier);
    final digest = sha256.convert(bytes);
    return _base64UrlEncode(digest.bytes);
  }

  /// Generate a random state parameter for CSRF protection.
  String _generateState() {
    final random = Random.secure();
    final bytes = List<int>.generate(32, (_) => random.nextInt(256));
    return _base64UrlEncode(bytes);
  }

  /// URL-safe base64 encoding without padding.
  String _base64UrlEncode(List<int> bytes) {
    return base64Url.encode(bytes).replaceAll('=', '');
  }

  /// Decode a base64url-encoded string to a UTF-8 string.
  String _decodeBase64Url(String input) {
    // Add padding if necessary.
    var padded = input;
    switch (padded.length % 4) {
      case 2:
        padded += '==';
        break;
      case 3:
        padded += '=';
        break;
    }
    final normalized = padded.replaceAll('-', '+').replaceAll('_', '/');
    return utf8.decode(base64.decode(normalized));
  }
}

File: adapters/frontend/kotlin/src/main/kotlin/com/rampart/RampartClient.kt
package com.rampart

import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import io.ktor.client.HttpClient
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.request.forms.submitForm
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.request
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpMethod
import io.ktor.http.contentType
import io.ktor.http.parameters
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.security.MessageDigest
import java.security.SecureRandom
import android.util.Base64

/**
 * Configuration for [RampartClient].
 *
 * @param issuer Rampart server URL (e.g. "https://auth.example.com").
 * @param clientId OAuth 2.0 client ID registered with the Rampart server.
 * @param redirectUri OAuth 2.0 redirect URI -- must match a registered URI.
 *   For Android, use a custom scheme (e.g. "com.example.app://callback").
 * @param scope OAuth 2.0 scopes (default: "openid").
 * @param httpClient Optional custom Ktor [HttpClient] for testing.
 */
data class RampartClientConfig(
    val issuer: String,
    val clientId: String,
    val redirectUri: String,
    val scope: String = "openid",
    val httpClient: HttpClient? = null,
)

/**
 * Token set returned by login and refresh operations.
 */
@Serializable
data class RampartTokens(
    @SerialName("access_token") val accessToken: String,
    @SerialName("refresh_token") val refreshToken: String,
    @SerialName("token_type") val tokenType: String = "Bearer",
    @SerialName("expires_in") val expiresIn: Int = 3600,
)

/**
 * Native Android authentication client for Rampart IAM.
 *
 * Implements OAuth 2.0 Authorization Code flow with PKCE using Chrome Custom Tabs
 * for the browser redirect. Tokens are stored in [EncryptedSharedPreferences].
 *
 * Usage:
 * ```kotlin
 * val client = RampartClient(context, RampartClientConfig(
 *     issuer = "https://auth.example.com",
 *     clientId = "my-android-app",
 *     redirectUri = "com.example.app://callback",
 * ))
 *
 * // Restore previous session
 * client.loadStoredTokens()
 *
 * // Start login
 * client.loginWithRedirect(activity)
 *
 * // In your Activity's onNewIntent / deep-link handler:
 * val tokens = client.handleCallback(intent.data!!)
 * val user = client.getUser()
 * ```
 */
class RampartClient(
    context: Context,
    config: RampartClientConfig,
) {
    private val issuer: String = config.issuer.trimEnd('/')
    private val clientId: String = config.clientId
    private val redirectUri: String = config.redirectUri
    private val scope: String = config.scope

    private val httpClient: HttpClient = config.httpClient ?: HttpClient(OkHttp)

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    private val prefs: SharedPreferences

    private var tokens: RampartTokens? = null

    init {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()

        prefs = EncryptedSharedPreferences.create(
            context,
            PREFS_FILE,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    // -------------------------------------------------------------------------
    // Public API
    // -------------------------------------------------------------------------

    /** Whether the client holds a non-expired access token. */
    val isAuthenticated: Boolean
        get() {
            val token = tokens?.accessToken ?: return false
            return try {
                val parts = token.split(".")
                if (parts.size != 3) return false
                val payload = decodeBase64Url(parts[1])
                val claims = json.decodeFromString<JwtExpClaims>(payload)
                val exp = claims.exp ?: return false
                (exp * 1000L) > System.currentTimeMillis()
            } catch (_: Exception) {
                false
            }
        }

    /** Returns the current access token, or null. */
    val accessToken: String?
        get() = tokens?.accessToken

    /** Returns the current token set, or null. */
    val currentTokens: RampartTokens?
        get() = tokens

    /**
     * Load tokens from encrypted storage. Call once at app startup to restore
     * a previous session.
     */
    fun loadStoredTokens() {
        val stored = prefs.getString(KEY_TOKENS, null) ?: return
        tokens = try {
            json.decodeFromString<RampartTokens>(stored)
        } catch (_: Exception) {
            prefs.edit().remove(KEY_TOKENS).apply()
            null
        }
    }

    /**
     * Start the OAuth login flow by opening the authorization URL in a
     * Chrome Custom Tab.
     *
     * After the user authenticates, the browser will redirect back to the app
     * via the custom URI scheme in [RampartClientConfig.redirectUri].
     * Handle the redirect in your Activity and call [handleCallback].
     *
     * @param context An Activity context used to launch the Custom Tab.
     */
    fun loginWithRedirect(context: Context) {
        val codeVerifier = generateCodeVerifier()
        val codeChallenge = generateCodeChallenge(codeVerifier)
        val state = generateState()

        // Persist PKCE values so they survive the browser redirect.
        prefs.edit()
            .putString(KEY_CODE_VERIFIER, codeVerifier)
            .putString(KEY_STATE, state)
            .apply()

        val authorizeUrl = Uri.parse("$issuer/oauth/authorize").buildUpon()
            .appendQueryParameter("response_type", "code")
            .appendQueryParameter("client_id", clientId)
            .appendQueryParameter("redirect_uri", redirectUri)
            .appendQueryParameter("scope", scope)
            .appendQueryParameter("state", state)
            .appendQueryParameter("code_challenge", codeChallenge)
            .appendQueryParameter("code_challenge_method", "S256")
            .build()

        val customTabsIntent = CustomTabsIntent.Builder().build()
        customTabsIntent.launchUrl(context, authorizeUrl)
    }

    /**
     * Handle the OAuth callback after the browser redirects back to the app.
     *
     * Pass the full callback [Uri] (including query parameters). This method
     * validates the state parameter, exchanges the authorization code for
     * tokens, and persists them in encrypted storage.
     *
     * @param uri The callback URI from the deep-link intent.
     * @return The token set received from the server.
     */
    suspend fun handleCallback(uri: Uri): RampartTokens {
        val code = uri.getQueryParameter("code")
        val state = uri.getQueryParameter("state")
        val error = uri.getQueryParameter("error")

        if (error != null) {
            val description = uri.getQueryParameter("error_description") ?: error
            throw RampartError.OAuthError(error, description)
        }

        if (code == null || state == null) {
            throw RampartError.InvalidCallback()
        }

        // Validate state.
        val storedState = prefs.getString(KEY_STATE, null)
        if (storedState == null || storedState != state) {
            cleanupPkceStorage()
            throw RampartError.StateMismatch()
        }

        val codeVerifier = prefs.getString(KEY_CODE_VERIFIER, null)
        if (codeVerifier == null) {
            cleanupPkceStorage()
            throw RampartError.MissingVerifier()
        }

        // Exchange code for tokens.
        val response: HttpResponse = httpClient.submitForm(
            url = "$issuer/oauth/token",
            formParameters = parameters {
                append("grant_type", "authorization_code")
                append("code", code)
                append("client_id", clientId)
                append("redirect_uri", redirectUri)
                append("code_verifier", codeVerifier)
            },
        )

        cleanupPkceStorage()

        val body = response.bodyAsText()
        if (response.status.value != 200) {
            throw RampartError.fromResponse(body, response.status.value)
        }

        val newTokens = json.decodeFromString<RampartTokens>(body)
        setTokens(newTokens)
        return newTokens
    }

    /**
     * Fetch the authenticated user profile from the `/me` endpoint.
     */
    suspend fun getUser(): RampartClaims {
        val response = authFetch("$issuer/me")
        val body = response.bodyAsText()
        if (response.status.value != 200) {
            throw RampartError.fromResponse(body, response.status.value)
        }
        return json.decodeFromString<RampartClaims>(body)
    }

    /**
     * Make an authenticated HTTP request with automatic Bearer token.
     *
     * On a 401 response, attempts one silent token refresh and retries.
     *
     * @param url The request URL.
     * @param method HTTP method (default: GET).
     * @param headers Additional headers.
     * @param body Optional request body (String).
     * @param contentType Content type for the body.
     */
    suspend fun authFetch(
        url: String,
        method: HttpMethod = HttpMethod.Get,
        headers: Map<String, String> = emptyMap(),
        body: String? = null,
        contentType: ContentType? = null,
    ): HttpResponse {
        suspend fun doFetch(): HttpResponse {
            return httpClient.request(url) {
                this.method = method
                header("Authorization", "Bearer ${tokens?.accessToken}")
                headers.forEach { (k, v) -> header(k, v) }
                if (body != null) {
                    contentType?.let { contentType(it) }
                    setBody(body)
                }
            }
        }

        var response = doFetch()

        if (response.status.value == 401 && tokens?.refreshToken != null) {
            try {
                refresh()
                response = doFetch()
            } catch (_: Exception) {
                // Refresh failed -- return the original 401.
            }
        }

        return response
    }

    /**
     * Refresh the access token using the stored refresh token.
     */
    suspend fun refresh(): RampartTokens {
        val refreshToken = tokens?.refreshToken
            ?: throw RampartError.NoRefreshToken()

        val response: HttpResponse = httpClient.request("$issuer/token/refresh") {
            method = HttpMethod.Post
            contentType(ContentType.Application.Json)
            setBody(json.encodeToString(RefreshRequest.serializer(), RefreshRequest(refreshToken)))
        }

        val body = response.bodyAsText()
        if (response.status.value != 200) {
            setTokens(null)
            throw RampartError.fromResponse(body, response.status.value)
        }

        val data = json.decodeFromString<RefreshResponse>(body)
        val updated = RampartTokens(
            accessToken = data.accessToken,
            refreshToken = refreshToken,
            tokenType = data.tokenType,
            expiresIn = data.expiresIn,
        )
        setTokens(updated)
        return updated
    }

    /**
     * Logout -- invalidates the refresh token on the server and clears local
     * tokens.
     */
    suspend fun logout() {
        val refreshToken = tokens?.refreshToken
        if (refreshToken != null) {
            try {
                httpClient.request("$issuer/logout") {
                    method = HttpMethod.Post
                    contentType(ContentType.Application.Json)
                    setBody(json.encodeToString(RefreshRequest.serializer(), RefreshRequest(refreshToken)))
                }
            } catch (_: Exception) {
                // Best-effort server logout.
            }
        }
        setTokens(null)
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    private fun setTokens(newTokens: RampartTokens?) {
        tokens = newTokens
        if (newTokens != null) {
            prefs.edit()
                .putString(KEY_TOKENS, json.encodeToString(RampartTokens.serializer(), newTokens))
                .apply()
        } else {
            prefs.edit().remove(KEY_TOKENS).apply()
        }
    }

    private fun cleanupPkceStorage() {
        prefs.edit()
            .remove(KEY_CODE_VERIFIER)
            .remove(KEY_STATE)
            .apply()
    }

    /** Generate a random 64-character code verifier (base64url-encoded). */
    private fun generateCodeVerifier(): String {
        val bytes = ByteArray(48)
        SecureRandom().nextBytes(bytes)
        return base64UrlEncode(bytes)
    }

    /** Compute S256 code challenge: BASE64URL(SHA256(verifier)). */
    private fun generateCodeChallenge(verifier: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(verifier.toByteArray(Charsets.UTF_8))
        return base64UrlEncode(digest)
    }

    /** Generate a random state parameter for CSRF protection. */
    private fun generateState(): String {
        val bytes = ByteArray(32)
        SecureRandom().nextBytes(bytes)
        return base64UrlEncode(bytes)
    }

    /** URL-safe base64 encoding without padding. */
    private fun base64UrlEncode(bytes: ByteArray): String {
        return Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
    }

    /** Decode a base64url-encoded string to a UTF-8 string. */
    private fun decodeBase64Url(input: String): String {
        val bytes = Base64.decode(input, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
        return String(bytes, Charsets.UTF_8)
    }

    companion object {
        private const val PREFS_FILE = "rampart_secure_prefs"
        private const val KEY_TOKENS = "rampart_tokens"
        private const val KEY_CODE_VERIFIER = "rampart_pkce_code_verifier"
        private const val KEY_STATE = "rampart_pkce_state"
    }
}

// ---------------------------------------------------------------------------
// Internal DTOs
// ---------------------------------------------------------------------------

@Serializable
internal data class JwtExpClaims(
    @SerialName("exp") val exp: Long? = null,
)

@Serializable
internal data class RefreshRequest(
    @SerialName("refresh_token") val refreshToken: String,
)

@Serializable
internal data class RefreshResponse(
    @SerialName("access_token") val accessToken: String,
    @SerialName("token_type") val tokenType: String = "Bearer",
    @SerialName("expires_in") val expiresIn: Int = 3600,
)
